### PR TITLE
FLUKA: Parsing of the projectile kinetic energy and amount of primary particles

### PR DESCRIPTION
**Pull request by lequ0n**

### CODE REVIEW

**Comment by grzanka** [comment]
@lequ0n fix the merge conflicts (rebase or merge master)

File: tests/fluka/test_fluka_converter.py [hunk 1]
```diff
@@ -67,15 +69,22 @@ def parser() -> Parser:
     """Parser fixture."""
     return get_parser_from_str("fluka")
 
+@pytest.fixture
```
Comment (grzanka) [inline tests/fluka/test_fluka_converter.py:72]: As for now I'd suggest to use the same JSON file as shieldhit,
take a look at TOPAS:
https://github.com/yaptide/converter/blob/114-parsing-of-the-projectile-kinetic-energy-fluka/tests/topas/conftest.py

Please define conftest.py in fluka tests suite. There define a fixture which provides Fluka JSON dict by passing a shieldhit dict.

At a later stage we will replace it with proper Fluka JSON (once the project will differentiate by editor)

File: tests/fluka/resources/project.json [hunk 2]
```diff
@@ -0,0 +1,644 @@
+{
```
Comment (grzanka) [inline tests/fluka/resources/project.json:1]: Remove, we should use shieldhit file

**Review by grzanka — CHANGES_REQUESTED** [review]

File: tests/fluka/test_fluka_converter.py [hunk 3]
```diff
@@ -7,7 +8,7 @@
 * default physics settings for hadron therapy
 DEFAULTS                                                              HADROTHE
 * beam source
-BEAM           -0.06       0.0       0.0      -2.0      -2.0          PROTON
+BEAM      1.5000E-01                                                  PROTON
```
Comment (grzanka) [inline tests/fluka/test_fluka_converter.py:11]: That seems wrong. 
Take a look here
https://flukafiles.web.cern.ch/manual/chapters/description_input/description_options/beam.html

positive number is interpreted as momentum, negative as energy

File: converter/fluka/input.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+from dataclasses import dataclass
+
+
+@dataclass
+class Input:
+    """Class mapping of the fluka input file."""
+
+    energy: float = 90.0  # GeV FLUKA specific
```
Comment (grzanka) [inline converter/fluka/input.py:8]: If you decide to keep energy in GeV, then I suggest to encode it in the parameter name, say `energy_GeV`

**Review by grzanka — CHANGES_REQUESTED** [review]

**Comment by lequ0n** [comment]
I've made changes to template formatting. We now have a class Card responsible for ensuring correct formatting, it's no longer done "manually".

File: converter/fluka/input.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+from dataclasses import dataclass
+from converter.fluka.card import Card
+
+
+@dataclass
+class Input:
+    """Class mapping of the fluka input file."""
+
+    energy_GeV: float = 90.0  # GeV FLUKA specific
```
Comment (grzanka) [inline converter/fluka/input.py:9]:
```suggestion
    energy_GeV: float = 0.15  # GeV FLUKA specific
```

This is a better default.

File: converter/fluka/card.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+@dataclass
+class Card:
+    """Class for a fluka card."""
+
+    tag: str = ""
+    what: Optional[List[str]] = None
+    sdum: str = ""
+
+    def __post_init__(self):
+        """Post init function."""
```
Comment (grzanka) [inline converter/fluka/card.py:14]:
```suggestion
        """Post init function, always called by automatically generated __init__"""
```

File: converter/fluka/card.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+from dataclasses import dataclass
+from typing import Optional, List
+
+
+@dataclass
+class Card:
+    """Class for a fluka card."""
+
+    tag: str = ""
+    what: Optional[List[str]] = None
```
Comment (grzanka) [inline converter/fluka/card.py:10]: Why don't you change type to `List[str]` and put a default value of `[]` ?

File: converter/fluka/card.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
+from dataclasses import dataclass
+from typing import Optional, List
+
+
+@dataclass
+class Card:
+    """Class for a fluka card."""
```
Comment (grzanka) [inline converter/fluka/card.py:7]:
```suggestion
    """Class representing one line (card) in Fluka input."""
```

File: converter/fluka/card.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+    tag: str = ""
+    what: Optional[List[str]] = None
+    sdum: str = ""
+
+    def __post_init__(self):
+        """Post init function."""
+        self.what = self.what or [""]
+        self.what += [""] * (6 - len(self.what))
```
Comment (grzanka) [inline converter/fluka/card.py:16]: Add a check if `len(self.what) > 6`

File: converter/fluka/card.py (hunk 7 above)
Comment (lequ0n) [reply converter/fluka/card.py:10]: I was receiving an error "mutable default for field is not allowed" so I changed to that. I must admit that at first I didn't see full error message which provided a solution, which I implemented now  `what: list[str] = field(default_factory=list)`

File: converter/fluka/card.py [hunk 10]
```diff
@@ -0,0 +1,29 @@
+from dataclasses import dataclass, field
+
+
+@dataclass
+class Card:
```
Comment (grzanka) [inline converter/fluka/card.py:5]: Add proper unit tests for this class.

**Review by grzanka — APPROVED** [review]
OK

**Review by grzanka — APPROVED** [review]
OK

**Pull request merged**